test/unit/hpa_validate_conf.c

```diff
@@ -0,0 +1,60 @@
+#include "test/jemalloc_test.h"
+
+static bool abort_called = false;
+static void (*default_malloc_message)(void *, const char *);
+
+static void
+mock_invalid_conf_abort(void) {
+	abort_called = true;
+}
+
+static void
+null_malloc_message(void *_1, const char* _2) {
+}
+
+TEST_BEGIN(test_hpa_validate_conf) {
+	test_skip_if(!hpa_supported());
+	void *ptr = malloc(4096);
+	/* Need to restore this here to see any possible assert messages */
+	malloc_message = default_malloc_message;
+	assert_true(abort_called,
+	     "Should have aborted due to invalid values for hpa_dirty_mult and "
+	     "hpa_hugification_treshold_ratio");
+	free(ptr);
+}
+TEST_END
+
+int
+main(void) {
+	/*
+	 * OK, this is a sort of nasty hack.  We don't want to add *another*
+	 * config option for HPA (the intent is that it becomes available on
+	 * more platforms over time, and we're trying to prune back config
+	 * options generally.  But we'll get initialization errors on other
+	 * platforms if we set hpa:true in the MALLOC_CONF (even if we set
+	 * abort_conf:false as well).  So we reach into the internals and set
+	 * them directly, but only if we know that we're actually going to do
+	 * something nontrivial in the tests.
+	 */
+	if (hpa_supported()) {
+		default_malloc_message = malloc_message;
+		malloc_message = null_malloc_message;
+		opt_hpa = true;
+		invalid_conf_abort = mock_invalid_conf_abort;
+		/*
+		 * We're forced to update the environment variable like this;
+		 * simply setting `opt_abort_conf = true` isn't enough, since
+		 * that might get overridden by the default compile-time malloc
+		 * conf settings.
+		 */
+		static char updated_malloc_conf[1025] =
```

Comment:
Why not setting abort_conf using MALLOC_CONF in hpa_validate_conf.sh ? That one overwrites the compile time options.

Reply:
IIRC (this was over a week ago), it's because if you set `abort_conf` in the shell-script then on FreeBSD (because Jemalloc is the system allocator) in CI instances where HPA isn't supported it will fail before it ever even launches the test.

Reply:
We should probably modify our testing setup so that we use a different prefix for the tests than we do for the system Jemalloc to avoid this kind of thing (it would also help cut down on all the warning messages you get about `prof:true` being an invalid conf value from the system allocator picking up test configs).

Reply:
I see. Yeah having a test only config would make sense. For the abort_conf here, maybe use the conf_harder hack? That overwrites everything and is process local. `https://github.com/jemalloc/jemalloc/blob/07a2eab3ed5dd76657ee689326acd9ecaf1e2830/src/jemalloc.c#L1075`

Reply:
Come to think of it, I think we can actually just use the default `malloc_conf` and not `malloc_conf_2_conf_harder`, since the only thing we're trying to override is the `--with-malloc-conf` default, which has the lowest priority of the various ways to set the conf anyway.